source/language/types.rst

```diff
@@ -271,64 +304,173 @@ be true and 0 will be false.
+initialization expression for a ``const`` declaration must be ``const S``, where
+``S`` is a type that is either ``T`` or can be :ref:`implicitly promoted
+<implicit-promotion-rules>` to ``T``.
+
+.. code-block::
+
+   // Valid statements
+
+   const uint u1 = 4;
+   const int[8] i1 = 8;
+   float[64] runtime_f1 = 2.0;
+
+   const uint u2 = u1;       // `u1` is of type `const uint`.
+   const float[32] f2 = u1;  // `const uint` is implicitly promoted to `const float[32]`.
+
+
+   // Invalid statements
+
+   const int[64] i2 = f2;  // `const float[32]` cannot be implicitly promoted to `const int[64]`.
+   const float[64] f3 = runtime_f1;  // `runtime_f1` is not `const`.
+
+
+.. _const-expression-operator:
+
+Operator expressions, e.g. ``a + b`` (addition), ``a[b]`` (bit-level indexing)
+and ``a == b`` (equality), and :ref:`certain built-in functions
+<const-expression-functions>` acting only on ``const`` operands will be
+evaluated at compile time.  The resulting values are of type ``const T``, where
+the type ``T`` is the type of the result when acting on non-\ ``const``
+operands.
+
+.. code-block::
+
+   // Valid statements
+
+   const uint[8] SIZE = 5;
 
-   // Declare a constant
-   const int my_const = 1234;
-   // Scientific notation is supported
-   const int[64] another_const = 1e12;
-   // Constant expressions are supported
-   const float[64] pi_by_2 = π / 2;
-   // Constants may be cast to real-time values
-   float[32] pi_by_2_val = float[32](pi_by_2)
+   const uint[16] u1 = 2 * SIZE;  // Compile-time value 10.
+   const float[64] f1 = 5.0 * SIZE;  // Compile-time value 25.0.
+   const bit b1 = u1[1];  // Compile-time value `"1"`.
+   const bit[SIZE - 1] b2 = u1[0:3];  // Compile-time value `"1010"`.
+
+
+.. _const-expression-cast:
+
+The resultant type of a cast to type ``T`` is ``const T`` if the input value has
+a type ``const S``, where values of type ``S`` can be cast to type ``T``.  If
+``S`` cannot be cast to ``T``, the expression is invalid.  The cast operator
+does not contain the keyword ``const``.
+
+.. code-block::
+
+   // Valid statements
+
+   const float[64] f1 = 2.5;
+   uint[8] runtime_u = 7;
+
+   const int[8] i1 = int[8](f1);  // `i1` has compile-time value 2.
+   const uint u1 = 2 * uint(f1);  // `u1` has compile-time value 4.
+
+
+   // Invalid statements
+
+   const bit[2] b1 = bit[2](f1);  // `float[64]` cannot be cast to `bit[2]`.
+   const int[16] i2 = int[16](runtime_u);  // Casting runtime values is not `const`.
+
+
+.. _const-expression-nonconst:
+
+The resultant type of any expression involving a value that is not ``const`` is
+not ``const``.  The output type of a call to a subroutine defined by a ``def``,
+or a call to a subroutine linked by an ``extern`` statement is not ``const``,
+although compilers are permitted to elide these calls at run-time if they can
+prove that doing so will not change the result of the program.  In these cases,
```

Comment:
**question**: I don't quite understand this clause. Mind explaining further?

Reply:
It might be better to remove that clause.  I'd meant that given
```qasm
def f(int a) -> int {
    return a;
}
```
a compiler would be allowed to transform `f(5)` into `5`, even through the non-`const` `def`.  When I wrote it, I was worried that the previous statement implied that that sort of classical optimisation (inlining, in this case) was forbidden.  Reading it again, I think that was an unfounded fear, and text about allowed classical transformations would go in a different part of the spec if we _were_ going to write about them anyway.

Reply:
I also read the language in this PR as not introducing any constraints on the optimizations a compiler may perform. So, removing this clause would be fine with me.

Reply:
I think leaving this clause in is reasonable.

Reply:
I think the statement itself is sound, but it's probably best to put it somewhere else - I think it fits better in a (currently non-existent) section about what classical optimisations and transformations compilers are permitted to make to the source code.  "Provably does not change the outcome of the program" is (I think) a pretty uncontroversial starting point, although there's definitely some nuance around not changing the outcome of a program that involves quantum randomness!